### PR TITLE
Rework DeviceGuard to restore original context upon the exit

### PR DESCRIPTION
**Pull request by JanuszL**
- some libraries are not using PrimiarContext while DALI does. When
  cudaSetDevice is called PrimaryContext is created and is set as
  the current one, the old one is lost while other apps may still need it.
 Adds saving of current context and restores it when DeviceGuard
 is destroyed
- reduced usage of CUContext as it is not needed in most cases

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

### CODE REVIEW

File: dali/core/device_guard.cc [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  old_context_(NULL) {
+  DALI_ENFORCE(cuInitChecked(),
+    "Failed to load libcuda.so. "
+    "Check your library paths and if the driver is installed correctly.");
+  CUDA_CALL(cuCtxGetCurrent(&old_context_));
+}
+
+DeviceGuard::DeviceGuard(int new_device) :
+  old_context_(NULL) {
+  if (new_device >= 0) {
```
Comment (mzient) [inline dali/core/device_guard.cc]: This check is wrong. If current context is not on device 0, DeviceGuard will not be able to use device 0.

File: dali/util/cucontext.h [hunk 2]
```diff
@@ -22,7 +22,7 @@ namespace dali {
 class CUContext {
  public:
   CUContext();
-  explicit CUContext(CUdevice device, unsigned int flags = 0);
+  explicit CUContext(int device_id_, unsigned int flags = 0);
```
Comment (mzient) [inline dali/util/cucontext.h]: I don't like it. You're mixing runtime and driver APIs at CUContext API level. It's very likely to cause trouble unless CUdevice and device ordinal from CUDA runtime are the same (I don't think they are).

File: dali/core/device_guard.cc (hunk 1 above)
Comment (JanuszL) [reply dali/core/device_guard.cc]: It is `>= 0` so we want to have noop for `-1` value, for example when tensor is on CPU, device assigned to it is `-1` so we want DeviceGuard to be noop and keep logic inside the buffer simple.

File: dali/util/cucontext.h (hunk 2 above)
Comment (JanuszL) [reply dali/util/cucontext.h]: Removing CUContext as it is no longer needed.

**Comment by JanuszL** [comment]
!build

File: dali/pipeline/data/buffer.h [hunk 3]
```diff
@@ -261,19 +264,9 @@ class Buffer {
...
   static void FreeMemory(void *ptr, size_t bytes, int device, bool pinned) {
-    // change to correct device for deletion
-    // Note: Can't use device guard due to potentially not GPUBackend.
-    int current_device = 0;
-    if (std::is_same<Backend, GPUBackend>::value) {
-      CUDA_CALL(cudaGetDevice(&current_device));
-      CUDA_CALL(cudaSetDevice(device));
-    }
+    // for device == -1 it is noop
+    DeviceGuard g(device);
```
Comment (mzient) [inline dali/pipeline/data/buffer.h]:
```suggestion
    if (std::is_same<Backend, CPUBackend>::value && !pinned) {
      DeviceGuard g(device);
      Backend::Delete(ptr, bytes, pinned);
    } else {
      Backend::Delete(ptr, bytes, pinned);
    }
```

File: dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h [hunk 4]
```diff
@@ -109,25 +111,32 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
...
-    for (int i = 0; i < num_threads_; ++i) {
-      NVJPEG_CALL(nvjpegBufferDeviceDestroy(device_buffer_[i]));
-      CUDA_CALL(cudaStreamDestroy(streams_[i]));
+      for (int i = 0; i < num_threads_; ++i) {
+        NVJPEG_CALL(nvjpegBufferDeviceDestroy(device_buffer_[i]));
+        CUDA_CALL(cudaStreamDestroy(streams_[i]));
+      }
+      CUDA_CALL(cudaEventDestroy(master_event_));
+      NVJPEG_CALL(nvjpegDestroy(handle_));
+    } catch (...) {
```
Comment (mzient) [inline dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: Everything we throw inherits from `std::exception`, so we can leverage it and print some meaningful diagnostic before aborting.
```suggestion
    } catch (const std::exception &e) {
      std::cerr << "Fatal error: exception in ~nvJPEGDecoder():\n" << e.what() << std::endl;
```

Comment (JanuszL) [reply dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: Done

File: dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h [hunk 5]
```diff
@@ -109,25 +111,32 @@ class nvJPEGDecoder : public Operator<MixedBackend>, CachedDecoderImpl {
...
+      // If destroying nvJPEG resources failed we are leaking something so terminate
+      std::terminate();
     }
-    CUDA_CALL(cudaEventDestroy(master_event_));
-    NVJPEG_CALL(nvjpegDestroy(handle_));
   }
+#pragma GCC diagnostic pop
```
Comment (mzient) [inline dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: Remove?

Comment (JanuszL) [reply dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h]: Done

File: include/dali/core/dynlink_cuda.h [hunk 6]
```diff
@@ -1886,9 +1886,9 @@ extern tcuMemcpy                       *cuMemcpy;
 extern tcuMemcpyPeer                   *cuMemcpyPeer;
 #endif
 
+extern tcuCtxCreate                    *cuCtxCreate;
```
Comment (mzient) [inline include/dali/core/dynlink_cuda.h]: Please move it to where other Ctx functions are.

Comment (JanuszL) [reply include/dali/core/dynlink_cuda.h]: Done

**Comment by JanuszL** [comment]
!build

File: dali/core/device_guard_test.cc [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+namespace dali {
+
+TEST(DeviceGuard, ConstructorWithDevice) {
+  int test_device = 0;
+  int guard_device = 0;
+  int current_device;
+  int count = 1;
+
+  cuInitChecked();
+  cudaGetDeviceCount(&count);
```
Comment (klecki) [inline dali/core/device_guard_test.cc]: Wouldn't it be beneficial to error check all the calls as well?
CUDA_CALL(cuda*) and CUDA_CALL(cu*)?

File: include/dali/core/device_guard.h [hunk 8]
```diff
@@ -17,38 +17,26 @@
...
+//  */
+class DLL_PUBLIC DeviceGuard {
+public:
   /// @brief Saves current device id and restores it upon object destruction
-  DeviceGuard() {
-    CUDA_CALL(cudaGetDevice(&original_device_));
-  }
+  DeviceGuard();
+
   /// @brief Saves current device id, sets a new one and switches back
```
Comment (klecki) [inline include/dali/core/device_guard.h]: You should write here that it's supposed to be no-op for -1

Comment (JanuszL) [reply include/dali/core/device_guard.h]: Done

File: dali/core/device_guard_test.cc (hunk 7 above)
Comment (JanuszL) [reply dali/core/device_guard_test.cc]: Done

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Pull request merged**